File: XWeather/iOS/ViewControllers/TableViewControllers/BaseTvc.cs
using System;
using System.Linq;

using Foundation;
using UIKit;

using XWeather.Clients;

namespace XWeather.iOS
{
	public class BaseTvc<TCell> : UITableViewController
		where TCell : BaseTvCell
	{

		internal nfloat FooterHeight = 44;


		public virtual nfloat HeaderHeight => 280;


		public WuLocation Location => WuClient.Shared.Selected;


		public BaseTvc (IntPtr handle) : base (handle) { }


		public override void ViewDidLoad ()
		{
			base.ViewDidLoad ();

			View.BackgroundColor = UIColor.Clear;

			TableView.BackgroundColor = UIColor.Clear;
		}


		[Export ("scrollViewDidScroll:")]
		public void Scrolled (UIScrollView scrollView)
		{
			MaskCells (scrollView);
		}


		public virtual void MaskCells (UIScrollView scrollView)
		{
			foreach (TCell cell in TableView.VisibleCells)
			{
				var topHiddenHeight = scrollView.ContentOffset.Y + HeaderHeight - cell.Frame.Y + scrollView.ContentInset.Top;
				var bottomHiddenHeight = cell.Frame.Bottom - (scrollView.ContentOffset.Y + scrollView.Frame.Height - FooterHeight);

				cell.SetTransparencyMask (topHiddenHeight, bottomHiddenHeight);
			}
		}


		public override nfloat GetHeightForHeader (UITableView tableView, nint section) => HeaderHeight;


		public override nfloat GetHeightForFooter (UITableView tableView, nint section) => FooterHeight;


		public override void WillDisplayHeaderView (UITableView tableView, UIView headerView, nint section)
		{
			if (TableView.VisibleCells.Any ()) MaskCells (TableView);

			var header = headerView as UITableViewHeaderFooterView;

			if (header?.ContentView != null)
				header.ContentView.BackgroundColor = UIColor.Clear;

			if (header?.BackgroundView != null)
				header.BackgroundView.BackgroundColor = UIColor.Clear;

			if (header?.TextLabel != null)
				header.TextLabel.TextColor = UIColor.White;
		}


		public override void WillDisplayFooterView (UITableView tableView, UIView footerView, nint section)
		{
			var footer = footerView as UITableViewHeaderFooterView;

			if (footer?.ContentView != null)
				footer.ContentView.BackgroundColor = UIColor.Clear;

			if (footer?.BackgroundView != null)
				footer.BackgroundView.BackgroundColor = UIColor.Clear;

			if (footer?.TextLabel != null)
				footer.TextLabel.TextColor = UIColor.White;
		}


		public TCell DequeCell (UITableView tableView, NSIndexPath indexPath)
			=> tableView.DequeueReusableCell (typeof (TCell).Name, indexPath) as TCell;


		public override UIStatusBarStyle PreferredStatusBarStyle () => UIStatusBarStyle.LightContent;
	}
}

File: Swift/XWeather/XWeather/ViewControllers/TableViewControllers/BaseTvc.swift
//
//  BaseTvc.swift
//  XWeather
//
//  Created by Colby Williams on 9/13/16.
//  Copyright © 2016 Colby Williams. All rights reserved.
//

import Foundation
import UIKit

class BaseTvc : UITableViewController {
	
	let headerHeight: CGFloat = 280
	let footerHeight: CGFloat = 44

	var cellId: String? {
		get {
			fatalError("subclasses of BaseTvc must override cellId")
		}
	}
	
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		view.backgroundColor = UIColor.clear
		
		tableView.backgroundColor = UIColor.clear
	}
	
	
	override func scrollViewDidScroll(_ scrollView: UIScrollView) {
		
		maskCells(scrollView)
	}
	
	
	func maskCells(_ scrollView: UIScrollView) {
		
		for cell in tableView.visibleCells {
			
			let topHiddenHeight = scrollView.contentOffset.y + headerHeight - cell.frame.minY + scrollView.contentInset.top
			
			let bottomHiddenHeight = cell.frame.maxY - (scrollView.contentOffset.y + scrollView.frame.height - footerHeight)
			
			cell.setTransparencyMask(top: topHiddenHeight, bottom: bottomHiddenHeight)
		}
	}
	
	
	override func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
		
		return headerHeight
	}
	
	
	override func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
		
		return footerHeight
	}
	
	
	override func tableView(_ tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
		
		if !tableView.visibleCells.isEmpty {
			
			maskCells(tableView)
		}
		
		if let header = view as? UITableViewHeaderFooterView {
			
			header.setBackgroundTransparent(labelColor: UIColor.white)
		}
	}
	
	
	override func tableView(_ tableView: UITableView, willDisplayFooterView view: UIView, forSection section: Int) {
		
		if let footer = view as? UITableViewHeaderFooterView {
			
			footer.setBackgroundTransparent(labelColor: UIColor.white)
		}
	}
	
	
	func dequeCell (_ tableView: UITableView, indexPath: IndexPath) -> UITableViewCell {
		
		return tableView.dequeueReusableCell(withIdentifier: cellId!, for: indexPath)
	}
	
	
	override var preferredStatusBarStyle: UIStatusBarStyle {
		
		return .lightContent
	}
}
